crates/oxc_query/src/properties.rs

```diff
@@ -0,0 +1,584 @@
+use std::convert::Into;
+
+use oxc_ast::ast::{BindingPatternKind, Expression, MemberExpression};
+use oxc_span::GetSpan;
+use trustfall::{
+    provider::{resolve_property_with, ContextIterator, ContextOutcomeIterator, ResolveInfo},
+    FieldValue,
+};
+
+use super::vertex::Vertex;
+use crate::{
+    util::{
+        accessibility_to_string, jsx_attribute_name_to_string, jsx_attribute_to_constant_string,
+        jsx_element_name_to_string,
+    },
+    vertex::InterfaceExtendVertex,
+    Adapter,
+};
+
+fn interface_extend_implem<'a, 'b: 'a>(
+    contexts: ContextIterator<'a, Vertex<'b>>,
+) -> ContextOutcomeIterator<'a, Vertex<'b>, FieldValue> {
+    resolve_property_with(contexts, |v| {
+        match v
+            .as_interface_extend()
+            .unwrap_or_else(|| {
+                panic!("expected to have an interfaceextend vertex, instead have: {v:#?}")
+            })
+            .as_ref()
+        {
+            InterfaceExtendVertex::Identifier(ident) => ident.name.to_string(),
+            InterfaceExtendVertex::MemberExpression(first_membexpr) => {
+                let MemberExpression::StaticMemberExpression(static_membexpr) = first_membexpr
+                else {
+                    unreachable!("TS:2499")
+                };
+                let mut parts = vec![static_membexpr.property.name.to_string()];
+                let mut membexpr = first_membexpr.object();
+                while let Expression::MemberExpression(expr) = membexpr {
+                    let MemberExpression::StaticMemberExpression(static_membexpr) = &expr.0 else {
+                        unreachable!("TS:2499")
```

Comment:
does this need more description, like what TS:2499 means?

Reply:
That'd be great, because this is a potentially user-facing error message.

It's also good practice to always include the data on the basis of which a panic was triggered. In this case, `expr`.